api/src/main/java/io/druid/data/input/impl/prefetch/RetryingInputStream.java

```diff
@@ -42,112 +51,125 @@
       T object,
       ObjectOpenFunction<T> objectOpenFunction,
       Predicate<Throwable> retryCondition,
-      int maxTry
+      int maxRetry
   ) throws IOException
   {
-    this.retryCondition = t -> {
-      Preconditions.checkNotNull(t);
+    this.object = object;
+    this.objectOpenFunction = objectOpenFunction;
+    this.retryCondition = retryCondition;
+    this.maxRetry = maxRetry;
+    this.delegate = new CountingInputStream(objectOpenFunction.open(object));
+  }
 
-      if (isConnectionReset(t)) {
+  private boolean isConnectionReset(Throwable t)
+  {
+    return (t instanceof SocketException && (t.getMessage() != null && t.getMessage().contains("Connection reset"))) ||
+           (t.getCause() != null && isConnectionReset(t.getCause()));
+  }
+
+  private void waitOrThrow(Throwable t, int i) throws IOException
+  {
+    final boolean isConnectionReset = isConnectionReset(t);
+    if (isConnectionReset || retryCondition.apply(t)) {
+      if (isConnectionReset) {
+        // Re-open the input stream on connection reset
         try {
           startOffset += delegate.getCount();
           log.info("retrying from offset[%d]", startOffset);
           delegate.close();
           delegate = new CountingInputStream(objectOpenFunction.open(object, startOffset));
-          return true;
         }
-        catch (IOException ioe) {
-          t.addSuppressed(ioe);
-          throw new RuntimeException(t);
+        catch (IOException e) {
+          t.addSuppressed(e);
+          throwAsIOException(t);
         }
       }
-
-      return retryCondition.apply(t);
-    };
-
-    this.maxTry = maxTry + 1;
-    this.delegate = new CountingInputStream(objectOpenFunction.open(object));
+      try {
+        RetryUtils.awaitNextRetry(t, null, i + 1, maxRetry, false);
+      }
+      catch (InterruptedException e) {
+        t.addSuppressed(e);
+        throwAsIOException(t);
+      }
+    } else {
+      throwAsIOException(t);
+    }
   }
 
-  private boolean isConnectionReset(Throwable t)
+  private void throwAsIOException(Throwable t) throws IOException
   {
-    return (t instanceof SocketException && (t.getMessage().contains("Connection reset"))) ||
-           (t.getCause() != null && isConnectionReset(t.getCause()));
+    Throwables.propagateIfInstanceOf(t, IOException.class);
+    throw new IOException(t);
   }
 
   @Override
   public int read() throws IOException
   {
-    try {
-      return RetryUtils.retry(
-          delegate::read,
-          retryCondition,
-          maxTry
-      );
-    }
-    catch (Exception e) {
-      throw new IOException(e);
+    for (int i = 0; i < maxRetry; i++) {
```

Comment:
Suggested to call the variable nTry, same below

Reply:
Done.